### PR TITLE
Task 150050512

### PR DESCRIPTION
**Pull request by alfonsomunozpomer**
Build passes tests:
- http://bar:8085/browse/ATLAS-BASECIDEVTASKS8
- http://bar:8085/browse/ATLAS-GXACIDEVTASKS7
- https://travis-ci.org/gxa/scxa-experiment-page-tsne-plot-view

If you want to actually give it a go start up gxa_sc and run `dev` script in `scxa-experiment-page-tsne-plot-view`.

### CODE REVIEW

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java [hunk 1]
```diff
@@ -27,7 +25,30 @@ public TSnePlotCollator(TSnePlotDao tSnePlotDao, ClusterDao clusterDao, GeneExpr
         this.geneExpressionDao = geneExpressionDao;
     }
 
-    public TreeMultimap<Integer, TSnePoint> getTSnePlotWithClusters(String experimentAccession, int k) {
+    public TreeMultimap<Integer, TSnePoint> getTsnePlotWithClustersAndExpression(String experimentAccession,
+                                                                                 int k,
+                                                                                 String geneId) {
+        TreeMultimap<Integer, TSnePoint> clusters = this.getTSnePlotWithClusters(experimentAccession, k);
```
Comment (wbazant) [inline sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java]: What's the purpose of `this.`?

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java [hunk 2]
```diff
@@ -26,7 +26,7 @@ public int compareTo(@Nullable TSnePoint o) {
         return o == null ? -1 : this.name().compareTo(o.name());
     }
 
-    // The public static method returning a TypeAdapter<Foo> is what tells auto-value-gson to create a TypeAdapter for Foo.
+    // Public static method returning a TypeAdapter<Foo> is what tells auto-value-gson to create a TypeAdapter for Foo
```
Comment (wbazant) [inline sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java]: If it works it's fine - looks fancy, and I'm not sure what it gives you. I just create toJson() methods everywhere.

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java [hunk 3]
```diff
@@ -7,14 +7,14 @@
 import javax.annotation.Nullable;
 
 @AutoValue
-public abstract class TSnePoint  implements Comparable<TSnePoint> {
+public abstract class   TSnePoint  implements Comparable<TSnePoint> {
```
Comment (wbazant) [inline sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java]: This is confusing, it's not clear from the class why it is `Comparable`, and the definition of comparing them by name doesn't correspond to any innate property of TSnePoints. I figured out that it's so you can put these in TreeMaps but there's no domain specific reason to do this.

File: sc/src/test/java/uk/ac/ebi/atlas/tsne/TSnePointTest.java [hunk 4]
```diff
@@ -15,6 +15,9 @@ public void create() throws Exception {
 
     @Test
     public void compareTo() throws Exception {
-        assertThat(TSnePoint.create(0.0, 0.0, 0.0, "B").compareTo(TSnePoint.create(1.0, 1.0, 1.0, "A")), is(greaterThan(0)));
+        assertThat(
```
Comment (wbazant) [inline sc/src/test/java/uk/ac/ebi/atlas/tsne/TSnePointTest.java]: This relies on people remembering how `compareTo` works and what it means when it returns a -1/0/1 . Possible improvement: make a list, sort it, and assert what the outcome is?

You could use this opportunity to explain the future developers why you want the TSnePoints compared by name.

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java [hunk 5]
```diff
@@ -27,7 +25,30 @@ public TSnePlotCollator(TSnePlotDao tSnePlotDao, ClusterDao clusterDao, GeneExpr
...
+        Map<String, Double> expression = geneExpressionDao.fetchGeneExpression(experimentAccession, geneId);
+
+        TreeMultimap<Integer, TSnePoint> clusterPointsWithExpression = TreeMultimap.create();
+        clusters.entries().forEach(
```
Comment (wbazant) [inline sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java]: Doesn't Java have `mapValues` ?

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java (hunk 1 above)
Comment (alfonsomunozpomer) [reply sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java]: Too much JavaScript... 😭

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java (hunk 5 above)
Comment (alfonsomunozpomer) [reply sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePlotCollator.java]: No, it doesn’t AFAIK. You can `collect(toMap(...))` but that returns a map, not a multimap.

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java (hunk 2 above)
Comment (alfonsomunozpomer) [reply sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java]: It’s a very simple way to add Gson (de)serializing support in `@AutoValue` classes. More here: https://github.com/rharter/auto-value-gson

File: sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java (hunk 3 above)
Comment (alfonsomunozpomer) [reply sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java]: 👍

Comment (alfonsomunozpomer) [reply sc/src/main/java/uk/ac/ebi/atlas/tsne/TSnePoint.java]: I’m a bit unsure if the comparator should be in `TSnePoint` or in `TSnePlotCollator`. I went with the first option to reuse it in `RandomTSnePlot`, but when this goes away we can move it to the collator.

**Pull request merged**